x-pack/filebeat/input/streaming/websocket.go

```diff
@@ -58,6 +65,47 @@ func (l *loggingRoundTripper) RoundTrip(req *http.Request) (*http.Response, erro
 	return resp, err
 }
 
+// heartBeat sends a ping message to the websocket server at regular intervals
+func (k *keepAlive) heartBeat(ctx context.Context, conn *websocket.Conn, start time.Time) context.CancelFunc {
+	ctx, cancel := context.WithCancel(ctx)
+	// set initial read deadline
+	conn.SetReadDeadline(time.Now().Add(k.cfg.KeepAlive.readControlDeadline))
+	// set pong handler to update read deadline
+	conn.SetPongHandler(func(string) error {
+		k.log.Debugw("received pong message from websocket server")
+		k.metrics.pongMessageReceivedTime.Update(time.Since(start).Nanoseconds())
+		return conn.SetReadDeadline(time.Now().Add(k.cfg.KeepAlive.readControlDeadline))
+	})
+
+	// set heartbeat ping routine
+	go func() {
+		ticker := time.NewTicker(k.cfg.KeepAlive.Interval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				if conn != nil {
+					err := conn.WriteControl(websocket.PingMessage, []byte{}, time.Now().Add(k.cfg.KeepAlive.WriteControlDeadline))
+					if err != nil {
+						k.log.Debugw("error sending ping message to websocket server:", err)
+						return
+					}
+					k.log.Debugw("sent ping message to websocket server")
+					k.metrics.pingMessageSendTime.Update(time.Since(start).Nanoseconds())
+				} else {
+					k.log.Debugw("websocket connection is nil but heartbeat routine is still running!")
```

Comment:
Does this happen? If it does, should we not return since `conn` is never going to be non-nil after this.

Reply:
This generally should ideally never happen, but I've been recently burned by gorilla/websocket libraries weird behaviour of returning nil response objects due to server crash/sudden disconnects, hence put this here for any such unexpected events.

Reply:
OK, so we should return then. Once `conn` is nil, there is no opportunity for it to become non-nil. In fact, the if should be hoisted out to the preamble for `(*keepAlive).heartBeat` since it is not possible for it to change its nilness from that point.

Reply:
Yes that could be done, but that would only provide safeguard against the initial trigger. What if the `conn` becomes `nil` when the routine is running because of some unexpected reason? This would cause the `conn.WriteControl` to panic. This was my thinking behind putting this check inside the select block.

Reply:
> What if the conn becomes nil when the routine is running because of some unexpected reason?

Can you show me a mechanism whereby that can happen that does not involve memory corruption?

Reply:
So I went through the whole control flow properly and yes, it seems this is unlikely to happen besides a in the case of a memory corruption. 

1. We only ever call `keepAlive.heartBeat()` if a connection is successfully established.
2. We always call heartBeatCancel() to gracefully stop this routine in all error/context cancellation scenarios.

So I'll remove this check and we don't even require a preamble check to exist, since we only trigger this on a successful connection.